File: Metal/FrontEnd/Parse/Grammar/Expression.cs
using Metal.FrontEnd.Scan;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Metal.FrontEnd.Parse.Grammar {
  abstract class Expression : AST<Expression> {

    internal interface IVisitor<T> {
      T Visit(Assign expression);
      T Visit(Literal expression);
      T Visit(Unary expression);
      T Visit(Binary expression);
      T Visit(Parenthesized expression);
      T Visit(Variable expression);
    }
    internal abstract T Accept<T>(IVisitor<T> visitor);

    internal class Assign : Expression {
      private Token name;
      private Expression value;

      internal override Expression Left => null;
      internal override Expression Right => null;

      internal override Token Operator => null;
      internal Token Name => name;
      internal Expression Value => value;

      internal Assign(Token name, Expression value) {
        this.name = name;
        this.value = value;
      }

      internal override T Accept<T>(IVisitor<T> visitor) {
        return visitor.Visit(this);
      }
    }

    internal class Unary : Expression {
      Token @operator;
      Expression right;
      internal Unary(Token @operator, Expression right) {
        this.@operator = @operator;
        this.right = right;
      }

      internal override T Accept<T>(IVisitor<T> visitor) {
        return visitor.Visit(this);
      }

      internal override Expression Left => null;
      internal override Expression Right => right;
      internal override Token Operator => @operator;
    }


    internal class Binary : Expression {
      private Expression left;
      private Token @operator;
      private Expression right;
      internal Binary(Expression left, Token @operator, Expression right) {
        this.left = left;
        this.@operator = @operator;
        this.right = right;
      }

      internal override T Accept<T>(IVisitor<T> visitor) {
        return visitor.Visit(this);
      }

      internal override Expression Left => null;
      internal override Expression Right => right;
      internal override Token Operator => @operator;
    }
    internal class Parenthesized : Expression {
      private Expression expression;
      internal Parenthesized(Expression expression) {
        this.expression = expression;
      }
      internal override T Accept<T>(IVisitor<T> visitor) {
        return visitor.Visit(this);
      }

      internal override Expression Left => null;
      internal Expression Center => expression;
      internal override Expression Right => null;
      internal override Token Operator => null;
    }

    internal class Literal : Expression {
      private object value;
      internal Literal(object value) {
        this.value = value;
      }

      internal override T Accept<T>(IVisitor<T> visitor) {
        return visitor.Visit(this);
      }
      internal override Expression Left => null;
      internal override Expression Right => null;
      internal override Token Operator => null;

      internal object Value => value;
    }

    internal class Variable : Expression {
      private Token name;
      internal Token Name => name;
      internal Variable(Token name) {
        this.name = name;
      }
      internal override Expression Left => null;

      internal override Expression Right => null;

      internal override Token Operator => null;

      internal override T Accept<T>(IVisitor<T> visitor) {
        return visitor.Visit(this);
      }
    }
  }
}

File: metal/frontend/parser/grammar/Expression.java
package metal.frontend.parser.grammar;

import java.util.List;
import metal.frontend.scanner.Token;

public abstract class Expression {
 public interface Visitor<R> {
    R visitBinaryExpression(Binary expression);
    R visitGroupExpression(Group expression);
    R visitLiteralExpression(Literal expression);
    R visitUnaryExpression(Unary expression);
  }

 public static class Binary extends Expression {
  public Binary(Expression left, Token operator, Expression right) {
      this.left = left;
      this.operator = operator;
      this.right = right;
    }

  public <R> R accept(Visitor<R> visitor) {
      return visitor.visitBinaryExpression(this);
  }

    public final Expression left;
    public final Token operator;
    public final Expression right;
  }

 public static class Group extends Expression {
  public Group(Expression expression) {
      this.expression = expression;
    }

  public <R> R accept(Visitor<R> visitor) {
      return visitor.visitGroupExpression(this);
  }

    public final Expression expression;
  }

 public static class Literal extends Expression {
  public Literal(Object value) {
      this.value = value;
    }

  public <R> R accept(Visitor<R> visitor) {
      return visitor.visitLiteralExpression(this);
  }

    public final Object value;
  }

 public static class Unary extends Expression {
  public Unary(Token operator, Expression right) {
      this.operator = operator;
      this.right = right;
    }

  public <R> R accept(Visitor<R> visitor) {
      return visitor.visitUnaryExpression(this);
  }

    public final Token operator;
    public final Expression right;
  }

 public abstract <R> R accept(Visitor<R> visitor);
}
